django/contrib/admin/sites.py

```diff
@@ -245,20 +245,17 @@ def wrapper(*args, **kwargs):
 
         # Add in each model's views, and create a list of valid URLS for the
         # app_index
-        valid_app_labels = []
         for model, model_admin in self._registry.items():
             urlpatterns += [
                 url(r'^%s/%s/' % (model._meta.app_label, model._meta.model_name), include(model_admin.urls)),
             ]
-            if model._meta.app_label not in valid_app_labels:
-                valid_app_labels.append(model._meta.app_label)
 
         # If there were ModelAdmins registered, we should have a list of app
         # labels for which we need to allow access to the app_index view,
-        if valid_app_labels:
-            regex = r'^(?P<app_label>' + '|'.join(valid_app_labels) + ')/$'
+        valid_app_labels = set(map(lambda x: x._meta.app_label, self._registry.keys()))
+        for app_label in valid_app_labels:
             urlpatterns += [
-                url(regex, wrap(self.app_index), name='app_list'),
+                url(r'^(?P<app_label>%s)/$' % app_label, wrap(self.app_index), name='app_list'),
```

Comment:
What exactly are you changing here? Looks as if we could completely revert the changes to this file?